File: binarySearch/searchInRotatedSortedArray.java
package binarySearch;
public class searchInRotatedSortedArray {
    public static void main(String[] args) {

        SolutionSISA solution = new SolutionSISA();

        System.out.println(solution.search(new int[] { 4, 5, 6, 7, 0, 1, 2 }, 0)); // 4
        System.out.println(solution.search(new int[] { 4, 5, 6, 7, 0, 1, 2 }, 3)); // -1
        System.out.println(solution.search(new int[] { 1, 3 }, 0)); // -1
        System.out.println(solution.search(new int[] { 1 }, 1)); // 0
        System.out.println(solution.search(new int[] { 1, 3 }, 1)); // 0
        System.out.println(solution.search(new int[] { 1, 3 }, 3)); // 1

    }
}

class SolutionSISA {
    public int search(int[] nums, int target) {
        int pivot = findPivot(nums);
        if(pivot ==-1) return binarySearch(nums, target,0, nums.length-1);
        if(target>=nums[0]) return binarySearch(nums,target, 0 , pivot);
        else return binarySearch(nums,target, pivot+1, nums.length-1);
        
    }

    public int binarySearch(int[] array , int target , int start , int end){
        while(start<=end){
            int mid = start +(end -start);
            if(array[mid]<target) start = mid +1;
            else if(array[mid]>target) end = mid-1;
            else return mid;
        }
        return -1;
    }
    
    public int findPivot(int[] array){
        int start = 0 ;
        int end = array.length -1 ;
        
        while(start <=end){
            int mid = start +(end - start) /2;
            //case 1 and case 2 
            if (mid >start && array[mid -1]> array [mid]) return mid -1;
            if (mid <end && array[mid]>array[mid+1]) return mid;
            
            //case 3
            if(array[start]<=array[mid]) start = mid +1;
            else end = mid -1 ;
        }
        return -1;
    }
}

File: search/searchInRotatedSortedArray.py
# https://leetcode.com/problems/search-in-rotated-sorted-array-ii/

class Solution(object):
    
    # Runtime: 64 ms, faster than 22.85% of Python online submissions for Search in Rotated Sorted Array II.
    # Memory Usage: 14 MB, less than 40.45% of Python online submissions for Search in Rotated Sorted Array II.
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: bool
        """
        def findPivot (nums):
            start  =0 
            end = len(nums)-1
            
            while start<= end :
                mid = start + (end-start)//2
                midElement = nums[mid]
                
                if ( mid<end and nums[mid]> nums[mid+1]): return mid
                if ( mid>start and nums[mid-1] > nums[mid]): return mid-1
                
                if (nums[start]== midElement and midElement == nums[end]):
                    
                    if (start<end and nums[start]>nums[start+1]): return start
                    start+=1
                    if (start < end and nums[end-1]>nums[end]) : return end-1
                    end-=1
                    
                elif (nums[start]< nums[mid] ) or (nums[start]==nums[mid]  and nums[mid]>nums[end]):
                    start=mid+1
                else :
                    end = mid-1
                    
                    
            return -1
        
        def binarySearch (nums, start, end ,target):
            while start <= end :
                mid = start+ (end -start)//2
                if nums[mid] == target : return True
                elif nums[mid]< target : start = mid+1
                else : end = mid-1
            return False
                
        
        
        pivot = findPivot(nums);
        if pivot != -1:
            if (target>= nums[0]):
                return binarySearch(nums, 0, pivot, target)
            else:
                return binarySearch(nums,pivot+1, len(nums)-1, target)
        else :
            return binarySearch (nums,0 , len(nums)-1, target)
        
